widget/gridwraplist.go

```diff
@@ -0,0 +1,355 @@
+	gwList := NewGridWrapList(
+		data.Length,
+		createItem,
+		func(i GridWrapListItemID, o fyne.CanvasObject) {
+			item, err := data.GetItem(int(i))
+			if err != nil {
+				fyne.LogError(fmt.Sprintf("Error getting data item %d", i), err)
+				return
+			}
+			updateItem(item, o)
+		})
+
+	data.AddListener(binding.NewDataListener(gwList.Refresh))
+	return gwList
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer.
+func (l *GridWrapList) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if f := l.CreateItem; f != nil {
+		if l.itemMin.IsZero() {
+			l.itemMin = f().MinSize()
+		}
+	}
+	layout := &fyne.Container{}
+	l.scroller = container.NewVScroll(layout)
+	layout.Layout = newGridWrapListLayout(l)
+	layout.Resize(layout.MinSize())
+	objects := []fyne.CanvasObject{l.scroller}
+	lr := newGridWrapListRenderer(objects, l, l.scroller, layout)
+	return lr
+}
+
+// MinSize returns the size that this widget should not shrink below.
+func (l *GridWrapList) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+func (l *GridWrapList) scrollTo(id GridWrapListItemID) {
+	if l.scroller == nil {
+		return
+	}
+	row := math.Floor(float64(id) / float64(l.getColCount()))
+	y := (float32(row) * l.itemMin.Height) + (float32(row) * theme.Padding())
+	if y < l.scroller.Offset.Y {
+		l.scroller.Offset.Y = y
+	} else if y+l.itemMin.Height > l.scroller.Offset.Y+l.scroller.Size().Height {
+		l.scroller.Offset.Y = y + l.itemMin.Height - l.scroller.Size().Height
+	}
+	l.offsetUpdated(l.scroller.Offset)
+}
+
+// Resize is called when this GridWrapList should change size. We refresh to ensure invisible items are drawn.
+func (l *GridWrapList) Resize(s fyne.Size) {
+	l.BaseWidget.Resize(s)
+	l.offsetUpdated(l.scroller.Offset)
+	l.scroller.Content.(*fyne.Container).Layout.(*gridWrapListLayout).updateList(true)
+}
+
+// ScrollTo scrolls to the item represented by id
+func (l *GridWrapList) ScrollTo(id GridWrapListItemID) {
+	length := 0
+	if f := l.Length; f != nil {
+		length = f()
+	}
+	if id < 0 || int(id) >= length {
+		return
+	}
+	l.scrollTo(id)
+	l.Refresh()
+}
+
+// ScrollToBottom scrolls to the end of the list
+func (l *GridWrapList) ScrollToBottom() {
+	length := 0
+	if f := l.Length; f != nil {
+		length = f()
+	}
+	if length > 0 {
+		length--
+	}
+	l.scrollTo(GridWrapListItemID(length))
+	l.Refresh()
+}
+
+// ScrollToTop scrolls to the start of the list
+func (l *GridWrapList) ScrollToTop() {
+	l.scrollTo(0)
+	l.Refresh()
+}
+
+// ScrollToOffset scrolls the list to the given offset position
+func (l *GridWrapList) ScrollToOffset(offset float32) {
+	l.scroller.Offset.Y = offset
+	l.offsetUpdated(l.scroller.Offset)
+}
+
+// GetScrollOffset returns the current scroll offset position
+func (l *GridWrapList) GetScrollOffset() float32 {
+	return l.offsetY
+}
```

Comment:
This seems a bit like internal details? What is the use case for them?

Reply:
I use it in my app to be able to reconstruct a new GridWrap(List) with the same data *and* at the same scroll position when the user navigates back to a previous page that had a grid view. In my history state I store the data model only and destroy the widgets. (And in the future to reduce allocations I'll reuse the widgets across different pages, just updating the backing data model, instead of destroying+creating new ones)

Reply:
That makes sense. I am fine with keeping it there in that case